README.md
# Treasure Data Import Tool in Java

## Overview

Many web/mobile applications generate huge amount of event logs (c,f. login,
logout, purchase, follow, etc).  Analyzing these event logs can be quite
valuable for improving services.  However, analyzing these logs easily and 
reliably is a challenging task.

Treasure Data Cloud solves the problem by having: easy installation, small 
footprint, plugins reliable buffering, log forwarding, the log analyzing, etc.

  * Treasure Data website: [http://treasure-data.com/](http://treasure-data.com/)
  * Treasure Data GitHub: [https://github.com/treasure-data/](https://github.com/treasure-data/)

**td-bulk-java** is a Java library, to access Treasure Data Cloud from Java application.

## Requirements

Java >= 1.6

## Install

### Install from GitHub repository

You can get latest source code using git.

    $ git clone https://github.com/treasure-data/td-import-java.git
    $ cd td-import-java
    $ mvn package

You will get the td-client jar file in td-import-java/target 
directory.  File name will be td-import-${td-import.version}-jar-with-dependencies.jar.
For more detail, see pom.xml.

## License

Apache License, Version 2.0



CHANGES.txt
Release 0.5.9 - 2015/08/04
  IMPROVEMENTS
    Improve parsing the URL of s3 source that includes s3 endpoint.
  BUG FIXED
    Changes parsing the URL of s3 source on Windows.

Release 0.5.8 - 2015/03/19
  IMPROVEMENTS
    Added new '--escape' option to 'import:prepare' command. default=backslash ('\')
  BUG FIXED
    Backslash '\' escape did not work for '\': '\\' is no longer parsed as '\\' but '\'.

Release 0.5.7 - 2015/02/26
  IMPROVEMENTS
    MySQLTableReader supported MEDIUMINT and MEDIUMINT UNSIGNED.
    Upgrades aws-java-sdk 1.7.5
    Added new '--empty-as-null-if-numeric' option to 'import:prepare' command

Release 0.5.6 - 2014/12/02
  IMPROVEMENTS
    MySQLTableReader supported the types of UNSIGNED TINYINT, UNSIGNED SMALLINT, UNSIGNED INT: it can translate those values to TreasureData type.

  BUG FIXED
    Changes the default value of retry-count: 8 to 10.

Release 0.5.5 - 2014/10/28
  IMPROVEMENTS
    Added validation for checking the status of perform job to waitPerform.

Release 0.5.4 - 2014/10/14
  IMPROVEMENTS
    Changed error-record file format to add line numbers of error records.
    Improved the behavior of time-value option: time-value should overwrite existing time colume values.

Release 0.5.3 - 2014/09/04
  BUG FIXED
    mysql: fixes bug: NPE occurs when Mysql table reader tries to create a sample record

Release 0.5.2 - 2014/07/29
  IMPROVEMENTS
    Added '--invalid-columns-handling' option to specify the behavior of the prepare process with regards to invalid
      column names. The 'warn' option is the default and causes the processor to emit a warning message when an invalid
      column name is encountered; the 'autofix' option makes the processor automatically fix the invalid column names by 
      making them lowercase and substituting invalid characters with underscores (_).

  BUG FIXED
    Fixes CharsetDecoder objects retrival by ThreadLocal: PrepareProcessors sometimes hit IllegalStateException.

Release 0.5.1 - 2014/07/23
  IMPROVEMENTS
    Updates td-client 0.5.1.

Release 0.5.0 - 2014/07/08
  IMPROVEMENTS
    Changes group ID for the artifact: 'com.treasure_data' to 'com.treasuredata'

Release 0.4.16 - 2014/07/02
  BUG FIXED
    When the columns specified with time-format are null or empty string, CSV/TSV reader throws NPE.

Release 0.4.15 - 2014/07/01
  IMPROVEMENTS
    Updates td-client 0.4.3.
    Improves error message by JSON file reader for boolean values.
    Improves error message by CSV file reader for type convertion failure.

  BUG FIXED
    Users with import_only access cannot import data by BulkImport.

Release 0.4.14 - 2014/06/17
  IMPROVEMENTS
    Updates td-client 0.4.2.

  BUG FIXED
    Fixes bug: import:auto from S3 doesn't work by causing upgrading td-client 0.4.2-SNAPSHOT.
    Fixes bug: import:auto runs without messages when no input files are given.

Release 0.4.13 - 2014/06/10
  IMPROVEMENTS
    Supports and can read from MySQL DATE, TIMESTAMP, and DATETIME columns.

Release 0.4.12 - 2014/04/23
  IMPROVEMENTS
    Changes default directory for outputting error-records: 'out' to 'error-records'
    Changes apache record reader.
    Upgrades td-client 0.4.0

  BUG FIXED
    Fixes bug: OOM error occurs when MySQLTableReader reads large table.

Release 0.4.11 - 2014/04/01
  IMPROVEMENTS
    Adds info. of td-client-java and td-import-java versions to log file.
    Adds time that each step takes to log file.

Release 0.4.10 - 2014/03/11
  IMPROVEMENTS
    Improves time-value option. It automatically allows generating and setting time values that are periodically sorted by hours specified by users.

Release 0.4.9 - 2014/03/04
  IMPROVEMENTS
    Changes JDK logging with append mode
    Changes upper limit of prepare-parallel option: 8 to 96

Release 0.4.8 - 2014/02/14
  IMPROVEMENTS
    Updates td-client 0.3.5.

Release 0.4.7 - 2014/02/04
  IMPROVEMENTS
    Changes behavior of auto-commit option: if a user executes upload command with auto-commit, he waits to finish commit process.

  BUG FIXED
    Fixes bug: prepare processor stops if it cannot parse invalid records during pre-execution (sampling) time.
    Fixes bug: JSON file reader cannot read and parse nil values.

Release 0.4.6 - 2013/12/20
  NEW FEATURES
    Supports item tables. To use bulk import for item tables, new 'primary-key' option is required.
    Adds retry-count to upload command.

  BUG FIXED
    Fixes bug: livelock occurs when upload command receives unexpected combinations of some options.
    Fixes bug: prepare processor never stops when EOFException occurs.

Release 0.4.5 - 2013/12/17
  NEW FEATURES
    Adds feature for outputing error records. Users can specify the directory as 'error-records-output'.

  BUG FIXED
    Fixes bug: NPE occurs when import:prepare is launched with non-existent file(s).

Release 0.4.4 - 2013/12/12
  NEW FEATURES
    Adds automatical type converter to 'mysql' table reader. It allows users to translate values of 'datetime' type into unix times of 'long' type automatically.

  IMPROVEMENTS
    Adds the feature for showing sample row to JSONFileReader class.
    Changes default quote character for TSV file reading: DOUBLE to NONE.
    Changes behavior for handling decoding errors like malformed inputs and unmappable chars. If you specify 'skip' as error-records-handling option, file reader replaces invalid char with '\uFFFD'. If you specify 'abort' as error-records-handling option, it aborts by invalid char.

Release 0.4.3 - 2013/12/06
  IMPROVEMENTS
    Implements MsgpackGZIPFileWriter#implemented write(TimeColumnValue, DoubleColumnValue). By the change, prepare command can accept unix time of both of long type and double type.

Release 0.4.2 - 2013/12/05
  NEW FEATURES
    Adds 'column-type' option to 'import:prepare' command. It allows users to specify a column type only instead of 'all-string' option.

  BUG FIXED
    Fixes bug: CSV/TSV file reader cannot parse columns that include '\"'.

src/main/java/org/apache/catalina/util/Strftime.java
/*
 * Copyright 1999,2004 The Apache Software Foundation.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.catalina.util;

import java.text.SimpleDateFormat;
import java.util.Properties;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Converts dates to strings using the same format specifiers as strftime
 *
 * Note: This does not mimic strftime perfectly.  Certain strftime commands, 
 *       are not supported, and will convert as if they were literals.
 *
 *       Certain complicated commands, like those dealing with the week of the year
 *       probably don't have exactly the same behavior as strftime.
 *
 *       These limitations are due to use SimpleDateTime.  If the conversion was done
 *       manually, all these limitations could be eliminated.
 *
 *       The interface looks like a subset of DateFormat.  Maybe someday someone will make this class
 *       extend DateFormat.
 *
 * @author Bip Thelin
 * @author Dan Sandberg
 * @version $Revision: 1.3 $, $Date: 2004/05/26 16:26:22 $
 */
public class Strftime {
    protected static Properties translate;
    protected SimpleDateFormat simpleDateFormat;

    /**
     * Initialize our pattern translation
     */
    static {
        translate = new Properties();
        translate.put("a","EEE");
        translate.put("A","EEEE");
        translate.put("b","MMM");
        translate.put("B","MMMM");
        translate.put("c","EEE MMM d HH:mm:ss yyyy");

        //There's no way to specify the century in SimpleDateFormat.  We don't want to hard-code
        //20 since this could be wrong for the pre-2000 files.
        //translate.put("C", "20");
        translate.put("d","dd");
        translate.put("D","MM/dd/yy");
        translate.put("e","dd"); //will show as '03' instead of ' 3'
        translate.put("F","yyyy-MM-dd");
        translate.put("g","yy");
        translate.put("G","yyyy");
        translate.put("H","HH");
        translate.put("h","MMM");
        translate.put("I","hh");
        translate.put("j","DDD");
        translate.put("k","HH"); //will show as '07' instead of ' 7'
        translate.put("l","hh"); //will show as '07' instead of ' 7'
        translate.put("m","MM");
        translate.put("M","mm");
        translate.put("n","\n");
        translate.put("p","a");
        translate.put("P","a");  //will show as pm instead of PM
        translate.put("r","hh:mm:ss a");
        translate.put("R","HH:mm");
        //There's no way to specify this with SimpleDateFormat
        //translate.put("s","seconds since ecpoch");
        translate.put("S","ss");
        translate.put("t","\t");
        translate.put("T","HH:mm:ss");
        //There's no way to specify this with SimpleDateFormat
        //translate.put("u","day of week ( 1-7 )");

        //There's no way to specify this with SimpleDateFormat
        //translate.put("U","week in year with first sunday as first day...");

        translate.put("V","ww"); //I'm not sure this is always exactly the same

        //There's no way to specify this with SimpleDateFormat
        //translate.put("W","week in year with first monday as first day...");

        //There's no way to specify this with SimpleDateFormat
        //translate.put("w","E");
        translate.put("X","HH:mm:ss");
        translate.put("x","MM/dd/yy");
        translate.put("y","yy");
        translate.put("Y","yyyy");
        translate.put("Z","z");
        translate.put("z","Z");
        translate.put("%","%");
    }


    /**
     * Create an instance of this date formatting class
     *
     * @see #Strftime( String, Locale )
     */
    public Strftime( String origFormat ) {
        String convertedFormat = convertDateFormat( origFormat );
        simpleDateFormat = new SimpleDateFormat( convertedFormat );
    }

    /**
     * Create an instance of this date formatting class
     * 
     * @param origFormat the strftime-style formatting string
     * @param locale the locale to use for locale-specific conversions
     */
    public Strftime( String origFormat, Locale locale ) {
        String convertedFormat = convertDateFormat( origFormat );
        simpleDateFormat = new SimpleDateFormat( convertedFormat, locale );
    }

    /**
     * Format the date according to the strftime-style string given in the constructor.
     *
     * @param date the date to format
     * @return the formatted date
     */
    public String format( Date date ) {
        return simpleDateFormat.format( date );
    }

    /**
     * Get the timezone used for formatting conversions
     *
     * @return the timezone
     */
    public TimeZone getTimeZone() {
        return simpleDateFormat.getTimeZone();
    }

    /**
     * Change the timezone used to format dates
     *
     * @see SimpleDateFormat#setTimeZone
     */
    public void setTimeZone( TimeZone timeZone ) {
        simpleDateFormat.setTimeZone( timeZone );
    }

    /**
     * Search the provided pattern and get the C standard
     * Date/Time formatting rules and convert them to the
     * Java equivalent.
     *
     * @param pattern The pattern to search
     * @return The modified pattern
     */
    protected String convertDateFormat( String pattern ) {
        boolean inside = false;
        boolean mark = false;
        boolean modifiedCommand = false;

        StringBuffer buf = new StringBuffer();

        for(int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);

            if ( c=='%' && !mark ) {
                mark=true;
            } else {
                if ( mark ) {
                    if ( modifiedCommand ) {
                        //don't do anything--we just wanted to skip a char
                        modifiedCommand = false;
                        mark = false;
                    } else {
                        inside = translateCommand( buf, pattern, i, inside );
                        //It's a modifier code
                        if ( c=='O' || c=='E' ) {
                            modifiedCommand = true;
                        } else {
                            mark=false;
                        }
                    }
                } else {
                    if ( !inside && c != ' ' ) {
                        //We start a literal, which we need to quote
                        buf.append("'");
                        inside = true;
                    }

                    buf.append(c);
                }
            }
        }

        if ( buf.length() > 0 ) {
            char lastChar = buf.charAt( buf.length() - 1 );

            if( lastChar!='\'' && inside ) {
                buf.append('\'');
            }
        }
        return buf.toString();
    }

    protected String quote( String str, boolean insideQuotes ) {
        String retVal = str;
        if ( !insideQuotes ) {
            retVal = '\'' + retVal + '\'';
        }
        return retVal;
    }

    /**
     * try to get the Java Date/Time formating associated with
     * the C standard provided
     *
     * @param c The C equivalent to translate
     * @return The Java formatting rule to use
     */
    protected boolean translateCommand( StringBuffer buf, String pattern, int index, boolean oldInside ) {
        char firstChar = pattern.charAt( index );
        boolean newInside = oldInside;

        //O and E are modifiers, they mean to present an alternative representation of the next char
        //we just handle the next char as if the O or E wasn't there
        if ( firstChar == 'O' || firstChar == 'E' ) {
            if ( index + 1 < pattern.length() ) {               
                newInside = translateCommand( buf, pattern, index + 1, oldInside );
            } else {
                buf.append( quote("%" + firstChar, oldInside ) );
            }
        } else {
            String command = translate.getProperty( String.valueOf( firstChar ) );

            //If we don't find a format, treat it as a literal--That's what apache does
            if ( command == null ) {
                buf.append( quote( "%" + firstChar, oldInside ) );
            } else {
                //If we were inside quotes, close the quotes
                if ( oldInside ) {
                    buf.append( '\'' );
                }
                buf.append( command );
                newInside = false;
            }
        }
        return newInside;
    }
}